### DISCUSSION TITLE
How did you know what to send to GoXLR ?

### DISCUSSION THREAD

**Discussion by JulanDeAlb** [oddbear/GoXLR-TouchPortal-Plugin#36] [Q&A] [answered]
Seeing your [docs](https://github.com/oddbear/GoXLR-TouchPortal-Plugin/wiki/Development---specification) how do you managed to know what you need to send in order to achieve something ?

 I know you have also a Loupedeck Plugin which is based on this one and since i got my Live S a few days ago i tried to code one myself with dynamic folder (more options to chose from how it should be displayed). I understand what you did but i dont get it how you know what specificly to send as json.
 
Like how did you know the RoutingAction are: "Turn On", "Turn Off", "Toggle" or did you just tried the values because they are named in StreamDeck like this ?

Best regards
Julian

**Comment by oddbear** [comment, marked as answer]
The absolute simplest way to see what is going on is to install Wireshark, run the Stream Deck plugin on a remote computer(/a virtual machine).
Then after running wireshark, connect from the GoXLR App (it would need the handshake to properly show the content).
Now filter on all traffic going between this two computers.

If you don't have two computers, you would need to create a filter that filters out all other things also.

There is also the possibility to create a "proxy", so the GoXLR App connects to some other application (you might need to create this one yourself), and that one connects to the plugin. Then you simply just log everything that goes through it.
However, with this approach, you would need two computers(/a virtual machine), as you are not allowed to run two "servers" with the same port on the same computer.

How this code works, is to "emulate" the Stream Deck plugin afterwards, so the GoXLR App thinks it's connected to a Stream Deck.
There are some bugs, or at least there used to be some between the GoXLR App and the Stream Deck, so be aware of those.

You will be limited however to what is implemented in the Stream Deck plugin, so routing, states and profiles. Unless you run the GoXLR on Linux where the Open Source drivers/utility can do basically everything through a custom API.

**Reply by JulanDeAlb** [reply]
Thank you very much for the quick response.
Ok this makes so sense.
It would be cool if they would release an API for the Windows Version or add more functionality to the StreamDeck plugin.

**Reply by oddbear** [reply]
I have tried the Linux version, and other interfaces like the Revelator series with unofficial API support, and can confirm that a API would be perfect for this device.

The way to maybe get this, is to send it in as an suggestion on the discord channel "feature-requests" on TC-Helicon Gaming. They do listen to the feedback there, but my guess is that 99% of the users does not know what an API is... and most of the requests could easily be fixed by a third party, if they had a proper API.

**Reply by JulanDeAlb** [reply]
So first of all, i made a new [Feature Request](https://discord.com/channels/456902246060392470/1043811166075490364/1043811166075490364) in the GoXLR Discord, requesting an API.

Second thing, i'm finished writing the GoXLR Plugin for the Loupedeck with the newest SDK and the ability for Dynamic Folders.
If you are interessted to have a look into it, just send me a quick response and i will add you to my private Repo (Mostly based/inspired by your plugin). Maybe we could work together to implement new features from the SDKs to add more Variants to use the Plugin if you wish (Since you already published it to the Marketplace, it doesnt make sense to add another one and because its based/inspired by your code)

**Reply by oddbear** [reply]
Nice. I would be very interesting to see what you have done. 😀
We could merge the code and release a new version. I think it will be possible to add you to the author list in the marked place to then.
